File: 백준/Silver/15650. N과 M （2）/N과 M （2）.java
import java.io.*;
import java.util.*;

public class Main {
	static int n, m;
	static boolean[] visited;
	static int[] sequence;
	static StringBuilder sb;
	
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine()); 
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        
       visited = new boolean[n+1];
       sequence = new int[m];
       sb = new StringBuilder();
       backtrack(0, 1);
       System.out.println(sb.toString());
    }
    
    private static void backtrack(int depth, int start) {
    	if(depth == m) {
    		for(int num : sequence) {
    			sb.append(num).append(" ");
    		}
    		sb.deleteCharAt(sb.length()-1);
    		sb.append("\n");
    		return;
    	}
    	
    	for(int i = start; i <= n; i++) {
    		if(!visited[i]) {
    			visited[i] = true;
    			sequence[depth] = i;
    			backtrack(depth+1, i+1);
    			visited[i] = false;
    		}
    	}
    }
    
  
    
    
}

File: 백준/Silver/15650. N과 M （2）/N과 M （2）.kt
package org.example

fun main() {
    val br = System.`in`.bufferedReader()
    val sb = StringBuilder()
    val (N, M) = br.readLine().split(" ").map { it.toInt() }
    // 1~N까지 자연수 중 중복 없이 M개 구하기
    val sequence = IntArray(M) // M개 고를 배열
    val visited = BooleanArray(N + 1) // 중복 없이, 0은 안 씀

    fun backtrack(depth: Int, start: Int) {
        if (depth == M) {
            sequence.forEach {
                sb.append(it).append(" ")
            }
            sb.deleteCharAt(sb.length - 1)
            sb.append("\n")
            return
        }

        for (i in start..N) {
            // 1~N까지의 수인데 고른 수열이 오름차순이 되어야 함.
            if (!visited[i]) {
                // 방문하지 않은 수라면
                visited[i] = true
                sequence[depth] = i
                backtrack(depth + 1, i + 1)
                visited[i] = false
            }
        }
    }
    backtrack(0, 1)
    print(sb)

    br.close()
}
